src/storage/src/source/postgres.rs

```diff
@@ -57,61 +57,60 @@ static WAL_LAG_GRACE_PERIOD: Duration = Duration::from_secs(5 * 60); // 5 minute
 static MAX_WAL_LAG: u64 = 100 * 1024 * 1024;
 
 trait ErrorExt {
-    fn is_recoverable(&self) -> bool;
+    fn is_definite(&self) -> bool;
 }
 
 impl ErrorExt for tokio::time::error::Elapsed {
-    fn is_recoverable(&self) -> bool {
-        true
+    fn is_definite(&self) -> bool {
+        false
     }
 }
 
 impl ErrorExt for tokio_postgres::Error {
-    fn is_recoverable(&self) -> bool {
+    fn is_definite(&self) -> bool {
         match self.source() {
-            Some(err) => {
-                match err.downcast_ref::<DbError>() {
-                    Some(db_err) => {
-                        use Severity::*;
-                        // Connection and non-fatal errors
-                        db_err.code() == &SqlState::CONNECTION_EXCEPTION
-                            || db_err.code() == &SqlState::CONNECTION_DOES_NOT_EXIST
-                            || db_err.code() == &SqlState::CONNECTION_FAILURE
-                            || db_err.code() == &SqlState::TOO_MANY_CONNECTIONS
-                            || db_err.code() == &SqlState::CANNOT_CONNECT_NOW
-                            || db_err.code() == &SqlState::ADMIN_SHUTDOWN
-                            || db_err.code() == &SqlState::CRASH_SHUTDOWN
-                            || !matches!(
-                                db_err.parsed_severity(),
-                                Some(Error) | Some(Fatal) | Some(Panic)
-                            )
+            Some(err) => match err.downcast_ref::<DbError>() {
+                Some(db_err) => {
+                    let class = &db_err.code().code()[0..2];
```

Comment:
(definitely deferring to postgres experts on this one)

Why is this approach (extra `.code()[0..2]`) preferable to the previous approach?  The three categories here seem to only have one error in each class?

That being said, if you'd like to keep the generality, what do you think of `let class = db_err.code().code().get(0..2)` where `class` is now an `Option`?  There's nothing that prevent `SqlState::code` from returning a `str` with too few characters -- so we could be defensive here so some garbage from an external system doesn't crash the process